### ISSUE TITLE
On the subject of "value*3"

### ISSUE THREAD

**Issue by jperrett256** [ChrisHouston/akai-midimix-reason-codec#2]
Hey there, I'm a C programmer trying to write some low-level software that interfaces with this hardware (AKAI MIDIMIX). Your code is the closest to documentation I have found for this purpose. (So, thanks!)

I am relatively new to MIDI, and I've never used Reason.
I managed to work out the inputs just fine, but the outputs have been bugging me.

From the MIDI spec, the Reason Remote codec documentation, and looking at your code, I gather that to turn on the button LEDs, I should send a message of the format:

1st byte: `0x90` ("Note On" command, channel 0)
2nd byte: e.g. `0x01` (code identifying the button whose LED we're trying to toggle)
3rd byte: `0x03` to enable, `0x00` to disable??

My question is about that third byte, which you set to `value*3`. Am I right to assume `value` would be 0 or 1, and that you are setting it to `0x03` to enable, `0x00` to disable?

I understand that a "Note On" command with a velocity of `0x00` is equivalent to a `Note Off`, and it makes sense that would translate to the context of button LEDs as well.
The part I don't understand is the enabling case: why it needs to be `0x03` and not simply `0x01` or something? Do you know what the extra bit encodes?

(What resources did you use to put this codec together, btw?)
Any help would be much appreciated.

**Comment by ChrisHouston** [comment]
It's been a while, but I seem to remember that setting the light values to "value" (i.e. the midi note volume that Reason sends out) wasn't high enough to turn on the lights, but *3 was. I can't remember if Reason sends a 0x01 value for boolean Ons, it might actually have been higher but still not sufficient to trigger the lights. Through trial and error I found that multiplying Reason's On value by 3 made it a high enough value for the Midimix.

I assume that Akai made the On value higher as a noise filter. Midi often has quite analoguey inputs - fingers lightly bumping keys, dirty switches not quite turning off, etc, so they didn't want the lights to be flickering from even the slightest signal.

As to the resources I used, it was code for another Akai device (good clues for how Akai sets up their midi devices), some Propellerhead/Reason Studios example code, and some midi inspection software - just a thing that displays the midi command on-screen when you press a button on the device.

**Comment by jperrett256** [comment]
Thanks for the information! I imagine that MIDI inspection software was more helpful in figuring out the inputs than the outputs. Looking for more codecs was a very helpful suggestion, though I had trouble finding any that configured outputs.

I was originally worried I would brick my device by sending it MIDI commands it wasn't expecting; obviously it should just ignore them, but I wanted to look/ask around before betting on that. Now I feel I have.
I have since done some experimenting, and thought I would share what I had found here, just in case anyone wanted it for whatever reason (probably only future me). I also happen to think it is very interesting. :)

As far as I can tell, if I send a Note On (`0x90`) command via [`midiOutShortMsg()`](https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/nf-mmeapi-midioutshortmsg), as long as the velocity (3rd byte in message) is in the range `0x01`-`0x7F`, it seems to turn the LED on. A velocity of `0x00` turns it off.

(Note I only tried velocity values of `0x00`, `0x01`, `0x02`, `0x03`, and `0x7F`. I'm extrapolating.)

Interestingly, a Note Off (`0x80`) command does not seem to do anything; turning off the LED _requires_ a Note On command with velocity set to `0x00`.

If anybody out there is looking for the "correct" answer for what velocity value to use: I ran Ableton Lite in a debugger, and found that it uses `0x7F` to turn LEDs on. (See RDX in the image below, which corresponds to the `dwMsg` parameter passed to [`midiOutShortMsg()`](https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/nf-mmeapi-midioutshortmsg).)

![x64dbg_ableton_midi_outshortmsg](https://user-images.githubusercontent.com/30437709/217645058-3496374b-ea77-411d-8b1e-c592f2d9ad84.png)

(Of course, Ableton uses a Note On command with velocity `0x00` to turn off the LEDs.)

All of this also means I have no idea why your Reason codec didn't work with "value" (as opposed to "value*3"), since any non-zero value enables LEDs for me, using the lowest-level MIDI API available on Windows (which Reason likely uses itself to implement its functionality). The explanation could be Reason specific, OS-specific, or (much less likely) a difference in the manufacturing of our devices (or something else entirely). I couldn't find anything helpful in the [Reason Remote codec documentation](https://developer.reasonstudios.com/downloads/other-products) (just says that for output, _"`value` is the new value of the control surface item"_, with no indication of what `value` might be).

I think at this point I will give up and just say I'm glad what you have works for you and what I have works for me.
